src/riak_core_apl.erl

```diff
@@ -59,62 +68,81 @@ active_owners(Ring, UpNodes) ->
     get_apl_chbin(DocIdx, N, CHBin, riak_core_node_watcher:nodes(Service)).
 
-%% Get the active preflist taking account of which nodes are up
-%% for a given chash/upnodes list
+%% @doc Get the active preflist taking account of which nodes are up
+%%      for a given chash/upnodes list.
 -spec get_apl_chbin(docidx(), n_val(), chashbin:chashbin(), [node()]) -> preflist().
 get_apl_chbin(DocIdx, N, CHBin, UpNodes) ->
     [{Partition, Node} || {{Partition, Node}, _Type} <-
                               get_apl_ann_chbin(DocIdx, N, CHBin, UpNodes)].
 
-%% Get the active preflist taking account of which nodes are up
-%% for a given ring/upnodes list
+%% @doc Get the active preflist taking account of which nodes are up
+%%      for a given ring/upnodes list.
 -spec get_apl(docidx(), n_val(), ring(), [node()]) -> preflist().
 get_apl(DocIdx, N, Ring, UpNodes) ->
-    [{Partition, Node} || {{Partition, Node}, _Type} <- 
+    [{Partition, Node} || {{Partition, Node}, _Type} <-
                               get_apl_ann(DocIdx, N, Ring, UpNodes)].
 
-%% Get the active preflist taking account of which nodes are up
-%% and annotate each node with type of primary/fallback
+%% @doc Get the active preflist taking account of which nodes are up for a given
+%%      chash/upnodes list and annotate each node with type of primary/fallback.
 get_apl_ann(DocIdx, N, UpNodes) ->
     {ok, CHBin} = riak_core_ring_manager:get_chash_bin(),
     get_apl_ann_chbin(DocIdx, N, CHBin, UpNodes).
 
-%% Get the active preflist taking account of which nodes are up
-%% for a given chash/upnodes list and annotate each node with type of
-%% primary/fallback
--spec get_apl_ann_chbin(binary(), n_val(), chashbin(), [node()]) -> preflist_ann().
-get_apl_ann_chbin(DocIdx, N, CHBin, UpNodes) ->
-    UpNodes1 = UpNodes,
-    Itr = chashbin:iterator(DocIdx, CHBin),
-    {Primaries, Itr2} = chashbin:itr_pop(N, Itr),
-    {Up, Pangs} = check_up(Primaries, UpNodes1, [], []),
-    Up ++ find_fallbacks_chbin(Pangs, Itr2, UpNodes1, []).
-
-%% Get the active preflist taking account of which nodes are up
-%% for a given ring/upnodes list and annotate each node with type of
-%% primary/fallback
+%% @doc Get the active preflist taking account of which nodes are up
+%%      for a given ring/upnodes list and annotate each node with type of
+%%      primary/fallback.
 -spec get_apl_ann(binary(), n_val(), ring(), [node()]) -> preflist_ann().
 get_apl_ann(DocIdx, N, Ring, UpNodes) ->
     UpNodes1 = UpNodes,
     Preflist = riak_core_ring:preflist(DocIdx, Ring),
-
     {Primaries, Fallbacks} = lists:split(N, Preflist),
     {Up, Pangs} = check_up(Primaries, UpNodes1, [], []),
     Up ++ find_fallbacks(Pangs, Fallbacks, UpNodes1, []).
 
-%% Same as get_apl, but returns only the primaries.
+%% @doc Get the active preflist for a given {bucket, key} and list of nodes
+%%      and annotate each node with type of primary/fallback.
+-spec get_apl_ann({binary(), binary()}, [node()]) -> preflist_ann().
```

Comment:
This spec will be wrong when considering bucket-types. Better to use `bucket()`, wherever it is defined.